t/lib/CRMS/PublicationDate.t

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/perl
+
+use strict;
+use warnings;
+
+use Test::Exception;
+use Test::More;
+
+use lib $ENV{'SDRROOT'} . '/crms/lib';
+use CRMS::PublicationDate;
+
+subtest '::new' => sub {
+  subtest 'without enumcron' => sub {
+    my $pd = CRMS::PublicationDate->new(date_type => 'm', date_1 => '1923', date_2 => '1963');
+    isa_ok($pd, 'CRMS::PublicationDate');
+  };
+
+  subtest 'with enumcron' => sub {
+    my $pd = CRMS::PublicationDate->new(date_type => 'm', date_1 => '1923', date_2 => '1963',
+      enumcron_date => '1923');
+    isa_ok($pd, 'CRMS::PublicationDate');
+  };
+};
+
+subtest '::clean_date' => sub {
+  is(CRMS::PublicationDate::clean_date('||||'), undef, 'no digits');
+  is(CRMS::PublicationDate::clean_date('1900'), '1900', 'all digits');
+  is(CRMS::PublicationDate::clean_date('19uu'), '19uu', 'partial date');
+  is(CRMS::PublicationDate::clean_date('0000'), undef, 'all-zero considered invalid');
+};
+
+subtest '::round_up' => sub {
+  is(CRMS::PublicationDate::round_up(), undef, 'undef');
+  is(CRMS::PublicationDate::round_up('1900'), '1900', 'fully-specified date');
+  is(CRMS::PublicationDate::round_up('190u'), '1909', 'unspecified year');
+  is(CRMS::PublicationDate::round_up('19uu'), '1999', 'unspecified decade');
+  is(CRMS::PublicationDate::round_up('uuuu', 1), '9999', 'completely unspecified in range');
+  is(CRMS::PublicationDate::round_up('uuuu'), undef, 'completely unspecified, no range');
+};
+
+subtest '::round_down' => sub {
+  is(CRMS::PublicationDate::round_down(), undef, 'undef');
+  is(CRMS::PublicationDate::round_down('1900'), '1900', 'fully-specified date');
+  is(CRMS::PublicationDate::round_down('190u'), '1900', 'unspecified year');
+  is(CRMS::PublicationDate::round_down('19uu'), '1900', 'unspecified decade');
+  is(CRMS::PublicationDate::round_down('uuuu'), '0000', 'completely unspecified');
+};
+
+subtest "#to_s" => sub {
+  my $pd = CRMS::PublicationDate->new(date_type => 'm', date_1 => '1923', date_2 => '1963');
+  is(ref $pd->to_s, '', '#to_s returns a string');
+};
+
+subtest "#format" => sub {
+  my $test_data = [
+    # type date1  date2   expected
```

Comment:
Love this comment and the way $test_data is laid out. Makes it very easy to see what's being tested.